src/m365/pp/commands/card/card-list.ts

```diff
@@ -0,0 +1,73 @@
+import { Logger } from '../../../../cli/Logger';
+import GlobalOptions from '../../../../GlobalOptions';
+import { odata } from '../../../../utils/odata';
+import { powerPlatform } from '../../../../utils/powerPlatform';
+import PowerPlatformCommand from '../../../base/PowerPlatformCommand';
+import commands from '../../commands';
+
+interface CommandArgs {
+  options: Options;
+}
+
+interface Options extends GlobalOptions {
+  environment: string;
+  asAdmin?: boolean;
+}
+
+class PpCardListCommand extends PowerPlatformCommand {
+  public get name(): string {
+    return commands.CARD_LIST;
+  }
+
+  public get description(): string {
+    return 'Lists Microsoft Power Platform cards in the specified Power Platform environment.';
+  }
+
+  public defaultProperties(): string[] | undefined {
+    return ['name', 'cardid', 'publishdate', 'createdon', 'modifiedon'];
+  }
+
+  constructor() {
+    super();
+
+    this.#initTelemetry();
+    this.#initOptions();
+  }
+
+  #initTelemetry(): void {
+    this.telemetry.push((args: CommandArgs) => {
+      Object.assign(this.telemetryProperties, {
+        asAdmin: !!args.options.asAdmin
+      });
+    });
+  }
+
+  #initOptions(): void {
+    this.options.unshift(
+      {
+        option: '-e, --environment <environment>'
+      },
+      {
+        option: '-a, --asAdmin'
+      }
+    );
+  }
+
+  public async commandAction(logger: Logger, args: any): Promise<void> {
+    if (this.verbose) {
+      logger.logToStderr(`Retrieving list of cards`);
+    }
+
+    const dynamicsApiUrl = await powerPlatform.getDynamicsInstanceApiUrl(args.options.environment, args.options.asAdmin);
```

Comment:
Moved this request in the `try` block in case it throws an error.

Reply:
But the error would be catched in the command base class if left out of the try catch, and render the appropriate error thrown from the method call right? I am not quite sure what moving it into the try catch solves in that case 🤔

Reply:
Agreed with @milanholemans that we should wrap it in a `try..catch` to ensure that we show a readable error. If the return type of this call is JSON, then we have util methods, like `handleRejectedOdataJsonPromise` and `handleRejectedJsonPromise` to get the error message out of the response in a user-friendly way.

Reply:
In that case I am still not sure if I understand.  `powerPlatform.getDynamicsInstanceApiUrl()` can throw an error as described https://github.com/pnp/cli-microsoft365/blob/main/src/utils/powerPlatform.ts#L29-L31 that is already a readable error, right?

Reply:
It's a readable error yes. But it's of type `Error` instead of `CommandError`, what will happen is that the `Error` object will be wrapped in a `CommandError` in the base class.
I changed the code a bit to ensure the command fails when retrieving the environment.
In the 1st example the code is executed as it is right now, in the 2nd example the retrieval of the environment is moved in the `try` block. As you can see, the first example starts with 2 times `Error: ` because we're wrapping an `Error` into a `CommandError`.

![image](https://user-images.githubusercontent.com/11723921/195997540-80b6b614-698d-4664-8d6e-937cc0914a0b.png)

Reply:
Thanks! Now it makes sense, couldn't wrap my head around it so sorry for buzzing you for more details 🤗